File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        int maxLength = 0;
        int st = 0;
        for(int i=0;i<s.length();i++){
           char c = s.charAt(i);
           while(st<s.length() && map.containsKey(c) && map.get(c) != 0){
            char a = s.charAt(st);
            map.put(a, map.get(a)-1);
            st++;
           }
           maxLength = Math.max(maxLength, i-st+1);
           map.put(c, 1);
           
        }

        return maxLength;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> set;
        int st = 0;
        int maxLength = 0;
        for(int rt=0;rt<s.length();rt++){
            char c = s[rt];

            while(set.contains(c)){
                set.erase(s[st]);
                st++;
            }

            maxLength = max(maxLength, rt-st+1);
            set.insert(c);
        }

        return maxLength;
    }
};